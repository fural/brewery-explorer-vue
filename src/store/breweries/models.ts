type BreweryType =
  | "micro"
  | "nano"
  | "regional"
  | "brewpub"
  | "large"
  | "planning"
  | "bar"
  | "contract"
  | "proprietor"
  | "closed";

export interface Brewery {
  id: string;
  name: string;
  brewery_type: BreweryType;
  street: string;
  address_2: string | null;
  address_3: string | null;
  city: string;
  state: string;
  county_province: string | null;
  postal_code: string;
  country: string;
  longitude: string | null;
  latitude: string | null;
  phone: string;
  website_url: string | null;
  updated_at: string;
  created_at: string;
}

export type GetBreweries = (params: BreweryFilters) => Promise<Brewery[]>;

export type BreweryApiFilters = {
  breweryName: string;
  perPage?: number;
};

export type BreweryUiFilters = {
  stateName: string;
};

export type BreweryFilters = BreweryApiFilters & BreweryUiFilters;

export interface State {
  breweries: Brewery[];
  filters: BreweryFilters;
  fetching: boolean;
  loading: boolean;
  error?: string;
}
