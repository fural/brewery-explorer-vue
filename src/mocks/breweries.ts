import { Brewery } from "../store/breweries/models";

export const breweries: Brewery[] = [
  {
    id: "13-stripes-brewery-taylors",
    name: "13 Stripes Brewery",
    brewery_type: "brewpub",
    street: "250 Mill St, Suite PW3101",
    address_2: null,
    address_3: null,
    city: "Taylors",
    state: "South Carolina",
    county_province: null,
    postal_code: "29687",
    country: "United States",
    longitude: null,
    latitude: null,
    phone: "8643491430",
    website_url: "http://www.13StripesBrewery.com",
    updated_at: "2023-02-21T01:41:59.300Z",
    created_at: "2023-02-21T01:41:59.300Z",
  },
  {
    id: "san-fernando-brewing-co-san-fernando",
    name: "San Fernando Brewing Co.",
    brewery_type: "micro",
    street: "425 Park Ave",
    address_2: null,
    address_3: null,
    city: "San Fernando",
    state: "California",
    county_province: null,
    postal_code: "91340-2525",
    country: "United States",
    longitude: "-118.4322067",
    latitude: "34.2828803",
    phone: "8187456175",
    website_url: "http://www.sanfernandobrewingcompany.com",
    updated_at: "2023-02-21T01:41:59.300Z",
    created_at: "2023-02-21T01:41:59.300Z",
  },
];
