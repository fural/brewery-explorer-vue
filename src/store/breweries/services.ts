import { api } from "../../utils/api";
import { Brewery, GetBreweries } from "./models";

const BASE_URL = "https://api.openbrewerydb.org";
const ENDPOINT = "breweries";
const PER_PAGE = 100;

export const getBreweries: GetBreweries = async (params) => {
  const url = new URL(`${BASE_URL}/${ENDPOINT}`);
  const { breweryName, perPage = PER_PAGE } = params;

  url.searchParams.append("per_page", String(perPage));

  if (breweryName) url.searchParams.append("by_name", breweryName);

  const response = await api.get<Brewery[]>(url.href);

  return response;
};
