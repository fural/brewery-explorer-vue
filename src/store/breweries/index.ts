import { reactive } from "vue";
import { useBreweries } from "./breweries";
import { State } from "./models";

const store = reactive<State>({
  breweries: [],
  filters: {
    breweryName: "",
    stateName: "",
    perPage: 100,
  },
  loading: true,
  fetching: false,
  error: undefined,
});

export { store, useBreweries };
