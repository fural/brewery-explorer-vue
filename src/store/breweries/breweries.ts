import { toRefs, watchEffect } from "vue";
import { store } from ".";
import { BreweryFilters } from "./models";
import { getBreweries } from "./services";

export const useBreweries = () => {
  const fetchBreweries = async (filters: BreweryFilters) => {
    try {
      store.fetching = true;
      const { stateName } = filters;
      const breweries = await getBreweries(filters);

      store.breweries = stateName
        ? breweries.filter((brewery) => brewery.state === stateName)
        : breweries;
    } catch (error) {
      store.error = "Something bad happened. Please try again.";
    } finally {
      store.fetching = false;
      store.loading = false;
    }
  };

  watchEffect(() =>
    fetchBreweries({
      breweryName: store.filters.breweryName,
      stateName: store.filters.stateName,
    })
  );

  const setSearchName = (name: string) => {
    store.filters.breweryName = name;
  };

  const resetFilters = () => {
    store.filters.breweryName = "";
    store.filters.stateName = "";
  };

  return {
    ...toRefs(store),
    setSearchName,
    resetFilters,
  };
};
