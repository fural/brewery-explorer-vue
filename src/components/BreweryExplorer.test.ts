import { render, fireEvent } from "@testing-library/vue";
import { breweries } from "../mocks/breweries";
import Component from "./BreweryExplorer.vue";

test("it should match snapshot", async () => {
  // Arrange
  const { container, findByText } = render(Component);

  // Assert
  expect(await findByText(breweries[0].name)).toBeTruthy();
  expect(container).toMatchSnapshot();
});

test("it should display the heading", async () => {
  // Arrange
  const { getByRole } = render(Component);
  const heading = getByRole("heading", { name: /Brewery Explorer/i });

  // Assert
  expect(heading).toBeTruthy();
});

test("it should filter by name", async () => {
  // Arrange
  const { getByRole, findByText } = render(Component);
  const input = getByRole("textbox", { name: "Filter by name" });

  // Act
  await fireEvent.update(input, "stripes");

  // Assert
  expect(input).toBeTruthy();
  expect(await findByText(breweries[0].name)).toBeTruthy();
});
