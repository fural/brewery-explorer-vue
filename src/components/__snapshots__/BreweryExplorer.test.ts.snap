// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it should match snapshot 1`] = `
<div>
  <div
    class="md:max-w-[1000px] m-auto py-4 md:px-8"
  >
    <div
      class="flex flex-col gap-2 px-4"
    >
      <div
        class="sticky top-0 flex flex-col gap-4"
      >
        <div
          class="flex items-center gap-2 text-3xl"
        >
          <span>
            üç∫
          </span>
          <h1
            class="font-bold text-slate-700"
          >
            Brewery Explorer
          </h1>
        </div>
        <div
          class="flex flex-col md:flex-row sm:items-center gap-4"
        >
          <label
            class="sm:w-3/5 relative"
          >
            <span
              class="sr-only"
            >
              Filter by name
            </span>
            <span
              class="absolute inset-y-0 left-0 flex items-center pl-2"
            >
              <svg
                class="h-5 w-5"
                fill="none"
                stroke="currentColor"
                strokeWidth="{1.5}"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                />
              </svg>
            </span>
            <input
              class="pl-9"
              name="Filter by name"
              placeholder="Filter by name"
              title="Filter by name"
              type="text"
            />
          </label>
          <div
            class="sm:w-1/3"
          >
            <select
              required=""
              title="Filter by state"
            >
              <option
                class="text-gray-400"
                value=""
              >
                Filter by state
              </option>
              
              <option
                value="Alabama"
              >
                Alabama
              </option>
              <option
                value="Alaska"
              >
                Alaska
              </option>
              <option
                value="Arizona"
              >
                Arizona
              </option>
              <option
                value="Arkansas"
              >
                Arkansas
              </option>
              <option
                value="California"
              >
                California
              </option>
              <option
                value="Colorado"
              >
                Colorado
              </option>
              <option
                value="Connecticut"
              >
                Connecticut
              </option>
              <option
                value="Delaware"
              >
                Delaware
              </option>
              <option
                value="Florida"
              >
                Florida
              </option>
              <option
                value="Georgia"
              >
                Georgia
              </option>
              <option
                value="Hawaii"
              >
                Hawaii
              </option>
              <option
                value="Idaho"
              >
                Idaho
              </option>
              <option
                value="Illinois"
              >
                Illinois
              </option>
              <option
                value="Indiana"
              >
                Indiana
              </option>
              <option
                value="Iowa"
              >
                Iowa
              </option>
              <option
                value="Kansas"
              >
                Kansas
              </option>
              <option
                value="Kentucky"
              >
                Kentucky
              </option>
              <option
                value="Louisiana"
              >
                Louisiana
              </option>
              <option
                value="Maine"
              >
                Maine
              </option>
              <option
                value="Maryland"
              >
                Maryland
              </option>
              <option
                value="Massachusetts"
              >
                Massachusetts
              </option>
              <option
                value="Michigan"
              >
                Michigan
              </option>
              <option
                value="Minnesota"
              >
                Minnesota
              </option>
              <option
                value="Mississippi"
              >
                Mississippi
              </option>
              <option
                value="Missouri"
              >
                Missouri
              </option>
              <option
                value="Montana"
              >
                Montana
              </option>
              <option
                value="Nebraska"
              >
                Nebraska
              </option>
              <option
                value="Nevada"
              >
                Nevada
              </option>
              <option
                value="New Hampshire"
              >
                New Hampshire
              </option>
              <option
                value="New Jersey"
              >
                New Jersey
              </option>
              <option
                value="New Mexico"
              >
                New Mexico
              </option>
              <option
                value="New York"
              >
                New York
              </option>
              <option
                value="North Carolina"
              >
                North Carolina
              </option>
              <option
                value="North Dakota"
              >
                North Dakota
              </option>
              <option
                value="Ohio"
              >
                Ohio
              </option>
              <option
                value="Oklahoma"
              >
                Oklahoma
              </option>
              <option
                value="Oregon"
              >
                Oregon
              </option>
              <option
                value="Pennsylvania"
              >
                Pennsylvania
              </option>
              <option
                value="Rhode Island"
              >
                Rhode Island
              </option>
              <option
                value="South Carolina"
              >
                South Carolina
              </option>
              <option
                value="South Dakota"
              >
                South Dakota
              </option>
              <option
                value="Tennessee"
              >
                Tennessee
              </option>
              <option
                value="Texas"
              >
                Texas
              </option>
              <option
                value="Utah"
              >
                Utah
              </option>
              <option
                value="Vermont"
              >
                Vermont
              </option>
              <option
                value="Virginia"
              >
                Virginia
              </option>
              <option
                value="Washington"
              >
                Washington
              </option>
              <option
                value="West Virginia"
              >
                West Virginia
              </option>
              <option
                value="Wisconsin"
              >
                Wisconsin
              </option>
              <option
                value="Wyoming"
              >
                Wyoming
              </option>
              
            </select>
          </div>
          <button
            class="bg-blue-800 hover:bg-blue-900 text-white font-bold py-2 px-4 rounded"
            type="reset"
          >
             Reset 
          </button>
        </div>
      </div>
      <div
        class="brewery__container scrollbar"
      >
        <div>
          <ul>
            
            <li
              class="brewery"
              data-testid="brewery-item"
            >
              <div
                class="brewery__title"
              >
                13 Stripes Brewery
              </div>
              <div
                class="brewery__state"
              >
                Taylors, South Carolina
              </div>
              <div
                class="brewery__link"
              >
                <div
                  class="w-7"
                >
                  <svg
                    class="w-5 h-5"
                    fill="none"
                    stroke="currentColor"
                    stroke-width="1.5"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20.893 13.393l-1.135-1.135a2.252 2.252 0 01-.421-.585l-1.08-2.16a.414.414 0 00-.663-.107.827.827 0 01-.812.21l-1.273-.363a.89.89 0 00-.738 1.595l.587.39c.59.395.674 1.23.172 1.732l-.2.2c-.212.212-.33.498-.33.796v.41c0 .409-.11.809-.32 1.158l-1.315 2.191a2.11 2.11 0 01-1.81 1.025 1.055 1.055 0 01-1.055-1.055v-1.172c0-.92-.56-1.747-1.414-2.089l-.655-.261a2.25 2.25 0 01-1.383-2.46l.007-.042a2.25 2.25 0 01.29-.787l.09-.15a2.25 2.25 0 012.37-1.048l1.178.236a1.125 1.125 0 001.302-.795l.208-.73a1.125 1.125 0 00-.578-1.315l-.665-.332-.091.091a2.25 2.25 0 01-1.591.659h-.18c-.249 0-.487.1-.662.274a.931.931 0 01-1.458-1.137l1.411-2.353a2.25 2.25 0 00.286-.76m11.928 9.869A9 9 0 008.965 3.525m11.928 9.868A9 9 0 118.965 3.525"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </div>
                <a
                  class="underline truncate max-w-[85%]"
                  href="http://www.13StripesBrewery.com"
                  target="_blank"
                >
                  http://www.13StripesBrewery.com
                </a>
              </div>
            </li>
            <li
              class="brewery"
              data-testid="brewery-item"
            >
              <div
                class="brewery__title"
              >
                San Fernando Brewing Co.
              </div>
              <div
                class="brewery__state"
              >
                San Fernando, California
              </div>
              <div
                class="brewery__link"
              >
                <div
                  class="w-7"
                >
                  <svg
                    class="w-5 h-5"
                    fill="none"
                    stroke="currentColor"
                    stroke-width="1.5"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20.893 13.393l-1.135-1.135a2.252 2.252 0 01-.421-.585l-1.08-2.16a.414.414 0 00-.663-.107.827.827 0 01-.812.21l-1.273-.363a.89.89 0 00-.738 1.595l.587.39c.59.395.674 1.23.172 1.732l-.2.2c-.212.212-.33.498-.33.796v.41c0 .409-.11.809-.32 1.158l-1.315 2.191a2.11 2.11 0 01-1.81 1.025 1.055 1.055 0 01-1.055-1.055v-1.172c0-.92-.56-1.747-1.414-2.089l-.655-.261a2.25 2.25 0 01-1.383-2.46l.007-.042a2.25 2.25 0 01.29-.787l.09-.15a2.25 2.25 0 012.37-1.048l1.178.236a1.125 1.125 0 001.302-.795l.208-.73a1.125 1.125 0 00-.578-1.315l-.665-.332-.091.091a2.25 2.25 0 01-1.591.659h-.18c-.249 0-.487.1-.662.274a.931.931 0 01-1.458-1.137l1.411-2.353a2.25 2.25 0 00.286-.76m11.928 9.869A9 9 0 008.965 3.525m11.928 9.868A9 9 0 118.965 3.525"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </div>
                <a
                  class="underline truncate max-w-[85%]"
                  href="http://www.sanfernandobrewingcompany.com"
                  target="_blank"
                >
                  http://www.sanfernandobrewingcompany.com
                </a>
              </div>
            </li>
            
          </ul>
        </div>
      </div>
      <div
        class="flex bg-blue-900 sticky bottom-0 h-16 mr-2 text-white text-sm"
      >
        <div
          class="flex flex-1 items-center p-4"
        >
          <div
            class="flex w-full justify-between text-center"
          >
            <div
              data-testid="numberOfResults"
            >
              Number of results: 2
            </div>
            <div>
              ¬© 2023 Brewery Explorer Ltd
            </div>
          </div>
        </div>
        <div
          class="flex items-center justify-center w-20"
        >
          <button
            class="flex items-center justify-center bg-blue-800 h-full w-full"
            data-testid="backToTopButton"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.5 10.5L12 3m0 0l7.5 7.5M12 3v18"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
